### PR TITLE
[REEF-1842]  Making IMRU task and input data association deterministic

### PR DESCRIPTION
**Pull request by jwang98052**
* Use partitionDescriptor id as part of the context id so that context id is associated with a fixed partitionDescriptor. Please notice that the partition descriptor id always follows the pattern like RandomInputPartition-x and is unique
* Make contexts stored in a sorted collection to ensure tasks with associated context/partitionDescriptor are always add to group in the same sequence.

JIRA: [REEF-1842](https://issues.apache.org/jira/browse/REEF-1842)
This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs [hunk 1]
```diff
@@ -392,6 +392,7 @@ private void SubmitTasks(IEnumerable<IActiveContext> activeContexts)
                         : GetMapperTaskIdByEvaluatorId(activeContext.EvaluatorId);
                     commGroup.AddTask(taskId);
                     taskIdAndContextMapping.Add(taskId, activeContext);
+                    Logger.Log(Level.Info, "Adding {0} with associated context: {1} to group.", taskId, activeContext.Id);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs]: We might also want to log which group we add the task to.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 2]
```diff
@@ -55,7 +56,7 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
         internal ServiceAndContextConfigurationProvider(IPartitionedInputDataSet dataset, ConfigurationManager configurationManager)
         {
             _dataset = dataset;
-            foreach (var descriptor in _dataset)
+            foreach (var descriptor in _dataset.Reverse())
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Why is this needed?

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Not necessary. It was to make input data description id and context id both follow ascending order, easy for debug. As I am changing the way to create context id, it is not needed anymore.

**Comment by jwang98052** [comment]
HashTable cannot be sorted. In fact, as long as we generate context ids and ensure they are unique and sorted by it, it should be safe.  I have made the change, removed the dependency on the partition data descriptor id for sorting.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 3]
```diff
@@ -138,32 +157,32 @@ internal ContextAndServiceConfiguration GetDataLoadingConfigurationForEvaluatorB
             }
 
             Logger.Log(Level.Info, "Getting a new data loading configuration");
-            _partitionIdProvider[evaluatorId] = _partitionDescriptorIds.Pop();
+            _partitionContextIdProvider[evaluatorId] = _avilablePartitionDescriptorContextIds.Pop();
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Wouldn't it be better to use
```C#
_partitionContextIdProvider.Add(evaluatorId, _avilablePartitionDescriptorContextIds.Pop());
```
and let it raise `ArgumentException` (thrown from `Add`) and `InvalidOperationException` (thrown from `Pop`) and then handle those exceptions instead of writing all those `if` statements upfront?

**Comment by markusweimer** [comment]
> Keep the lines' length under 120, or, better yet, 100 characters.

@motus, the style guide prescribes 120 characters. Let's keep it that way :)

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 4]
```diff
@@ -42,22 +42,41 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
...
+
+        /// <summary>_configurationManager
+        /// Configuration manager that provides configurations
+        /// </summary>
         private readonly ConfigurationManager _configurationManager;
 
         /// <summary>
-        /// Constructs the object witch maintains partitionDescriptor Ids so that to provide proper data load configuration 
+        /// Constructs the object which maintains partitionDescriptor Ids so that to provide proper data load configuration
+        /// It also maintain the partitionDescriptor id and context id mapping to ensure same context id alway assign the same data partition
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: maintains
maintains  
No difference :)

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 5]
```diff
@@ -42,22 +42,41 @@ internal sealed class ServiceAndContextConfigurationProvider<TMapInput, TMapOutp
...
-        private readonly Stack<string> _partitionDescriptorIds = new Stack<string>();
+        /// <summary>
+        /// Mapping between Evaluator id and assigned partition descriptor/context ids
+        /// </summary>
+        private readonly Dictionary<string, PartitionDescriptorContextIdBundle> _partitionContextIdProvider = new Dictionary<string, PartitionDescriptorContextIdBundle>();
+
+        /// <summary>
+        /// Available partition descriptor and context ids stack
+        /// </summary>
+        private readonly Stack<PartitionDescriptorContextIdBundle> _avilablePartitionDescriptorContextIds = new Stack<PartitionDescriptorContextIdBundle>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Done

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 6]
```diff
@@ -70,13 +89,13 @@ internal ServiceAndContextConfigurationProvider(IPartitionedInputDataSet dataset
         /// <returns>Whether failed evaluator is master or not</returns>
         internal void RemoveEvaluatorIdFromPartitionIdProvider(string evaluatorId)
         {
-            if (!_partitionIdProvider.ContainsKey(evaluatorId))
+            if (!_partitionContextIdProvider.ContainsKey(evaluatorId))
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Here the purpose is not to get value but a validation checking to make sure there is no existing key there.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 7]
```diff
@@ -105,29 +124,29 @@ internal ContextAndServiceConfiguration GetContextConfigurationForMasterEvaluato
         /// <returns></returns>
         internal string GetPartitionIdByEvaluatorId(string evaluatorId)
         {
-            if (!_partitionIdProvider.ContainsKey(evaluatorId))
+            if (!_partitionContextIdProvider.ContainsKey(evaluatorId))
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Done

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Done

**Comment by jwang98052** [comment]
@motus @markusweimer I have addressed your review comments. Please have another look. Thanks!

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 6 above)
Comment (motus) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: I am talking about the entire method. You make three trips to the same entry in the Dictionary: `_partitionContextIdProvider.ContainsKey(evaluatorId)`, then `_partitionContextIdProvider[evaluatorId]`, and, finally, `_partitionContextIdProvider.Remove(evaluatorId)`. My point is that you can do it with fewer calls, e.g.
```C#
PartitionDescriptorContextIdBundle partition;
if(_partitionContextIdProvider.TryGetValue(evaluatorId, out partition))
{
    _availablePartitionDescriptorContextIds.Push(partition);
    _partitionContextIdProvider.Remove(evaluatorId);
}
// ...
```
i.e. at least there is no `[]`. (I wish C# `.Remove()` returned the vaule, and not bool)

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 4 above)
Comment (motus) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: As in: *It also maintain**s** the partitionDescriptor id...*

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Done

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Good point!

**Comment by jwang98052** [comment]
@motus Have addressed your review comments. Thanks!

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 8]
```diff
@@ -92,14 +92,19 @@ internal ServiceAndContextConfigurationProvider(IPartitionedInputDataSet dataset
         /// <returns>Whether failed evaluator is master or not</returns>
         internal void RemoveEvaluatorIdFromPartitionIdProvider(string evaluatorId)
         {
-            if (!_partitionContextIdProvider.ContainsKey(evaluatorId))
+            PartitionDescriptorContextIdBundle partitionDescriptor;
+            if (_partitionContextIdProvider.TryGetValue(evaluatorId, out partitionDescriptor))
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: instead of lines 95-96 you can simply do 
```C#
if (_partitionContextIdProvider.TryGetValue(evaluatorId, out var partitionDescriptor))
```

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs [hunk 9]
```diff
@@ -92,14 +92,19 @@ internal ServiceAndContextConfigurationProvider(IPartitionedInputDataSet dataset
...
-                Exceptions.Throw(new Exception(msg), Logger);
+                _availablePartitionDescriptorContextIds.Push(partitionDescriptor);
+                _partitionContextIdProvider.Remove(evaluatorId);
+            }
+            else
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture,
+                    "Partition descriptor for Failed evaluator:{0} not present",
+                    evaluatorId);
+                throw new Exception(msg);
```
Comment (motus) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Can we throw some more specific exception here?

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Doesn't work for me. Do you use some newer version of .Net?

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs]: Sure. Change it to IMRUSystemException.

**Comment by jwang98052** [comment]
@motus Have addressed your review comments and pushed the change.
